src/repository/NodeRepository.ts

```diff
@@ -0,0 +1,63 @@
+import { Logger } from '@streamr/utils'
+import { RowDataPacket } from 'mysql2'
+import { Inject, Service } from 'typedi'
+import { Topology } from '../crawler/Topology'
+import { Nodes } from '../entities/Node'
+import { createSqlQuery } from '../utils'
+import { ConnectionPool } from './ConnectionPool'
+
+interface NodeRow extends RowDataPacket {
+    id: string
+    ipAddress: string | null
+}
+
+const logger = new Logger(module)
+
+@Service()
+export class NodeRepository {
+
+    private readonly connectionPool: ConnectionPool
+
+    constructor(
+        @Inject() connectionPool: ConnectionPool
+    ) {
+        this.connectionPool = connectionPool
+    }
+
+    async getNodes(
+        ids?: string[],
+        pageSize?: number,
+        cursor?: string
+    ): Promise<Nodes> {
+        logger.info('Query: getNodes', { ids, pageSize, cursor })
+        const whereClauses = []
+        const params = []
+        if (ids !== undefined) {
+            whereClauses.push('id in (?)')
+            params.push(ids)
+        }
+        const sql = createSqlQuery(
+            `SELECT id, ipAddress FROM nodes`,
+            whereClauses
+        )
+        return this.connectionPool.queryPaginated<NodeRow[]>(sql, params)
+    }
+
+    async replaceNetworkTopology(topology: Topology): Promise<void> {
+        const nodes = topology.getNodes().map((node) => {
+            return [node.id, node.ipAddress]
```

Comment:
Can field `node.ipAddress` be undefined here? Does the insertion still work properly (i.e. library handles the conversion from undefined => NULL)

Reply:
Checked, works ok.